### PR TITLE
Added BuildInformationUploader

### PR DESCRIPTION
**Pull request by rain-on**
In order to contain hypermedia linkage (eg rootDoc --> Space --> Project --> xxx) within The Java SDK (as opposed to leaking the nested HTTP calls to a client of the library) - the concept of a BuildInformationUploader has been created.

This class encapsulates the hypermedia linkage, and ultimately, will allow uses to use either a "direct" API (when available) or traverse the hypermedia links.

This pattern is to be repeated for all resources.

### CODE REVIEW

File: octopus-sdk/src/test/java/com/octopus/sdk/operations/buildinformation/OctopusBuildInformationUploaderTest.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    assertThat(argCaptor.getAllValues()).hasSize(1);
+    assertThat(argCaptor.getValue().getPath()).isEqualTo("/api/arbitrary-path");
+    assertThat(argCaptor.getValue().getQueryParameters().keySet()).containsExactly("overwriteMode");
+    assertThat(argCaptor.getValue().getQueryParameters().values())
+        .containsExactly(singletonList(OverwriteMode.OverwriteExisting.toString()));
+  }
+
+  @Test
+  public void exceptionIsThrownIfSpaceNotSpecifiedAndDefaultSpaceNotSupported()
+      throws MalformedURLException, URISyntaxException {
```
Comment (long-sean) [inline octopus-sdk/src/test/java/com/octopus/sdk/operations/buildinformation/OctopusBuildInformationUploaderTest.java:103]: Cannot see anything in the call chain that could result in URISyntaxException?

**Review by long-sean — COMMENTED** [review]
Just one potentially unused exception type in test method signature

File: octopus-sdk/src/main/java/com/octopus/sdk/operations/buildinformation/BuildInformationUploaderContextBuilder.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    return this;
+  }
+
+  public BuildInformationUploaderContext build() throws MalformedURLException, URISyntaxException {
+    Preconditions.checkNotNull(
+        packageVersion, "packageVersion must be set on a build information object");
+    Preconditions.checkNotNull(
+        overwriteMode, "overwriteMode must be set on a build information object");
+    Preconditions.checkNotNull(
+        buildUrl, "teamCityServerUrl must be set on a build information object");
```
Comment (long-sean) [inline octopus-sdk/src/main/java/com/octopus/sdk/operations/buildinformation/BuildInformationUploaderContextBuilder.java:113]: SDK specifically tied to TeamCity?

**Review by long-sean — APPROVED** [review]

File: octopus-sdk/src/main/java/com/octopus/sdk/operations/buildinformation/BuildInformationUploaderContextBuilder.java (hunk 2 above)
Comment (rain-on) [reply octopus-sdk/src/main/java/com/octopus/sdk/operations/buildinformation/BuildInformationUploaderContextBuilder.java:113]: whoops - fixed.

**Review by long-sean — APPROVED** [review]

**Pull request merged**